### ISSUE TITLE
FR: Use real bold CJK font in UI with font subsetting

### ISSUE THREAD

**Issue by weijiuqiao** [koreader/koreader#9226]
Currently bold CJK characters in UI are synthetized and can appear less ideal especially when there are more strokes. I presume the lack of a real bold CJK font is due to its huge file size. If that's the case, we can make a subsetted font with only characters needed in UI display.

One available tool [pyftsubset](https://fonttools.readthedocs.io/en/latest/subset/index.html) makes subsetting a font really easy. I tried subsetting a NotoCJK Bold font (17MB) to one with all characters found in the related translation files (under ja, ko_KR, zh_CN and zh_TW folders) and the result is 872k in size. It's still very much bloated since most of the extracted characters never show in bold. 

I don't think we can filter out the actual bold ones programmatically though (hope I'm wrong), so this would mean an increase of around 1MB in total size. Also if chosen as the document font, it results in real and fake bold mixed up. The good thing though is that pyftsubset keeps the locl feature intact which is essential in our use case.

**Comment by poire-z** [comment]
You could see how it looks with a less strong synthesized bold:
```diff
--- a/frontend/ui/font.lua
+++ b/frontend/ui/font.lua
@@ -157,12 +157,18 @@ end
 -- local bold_strength_factor = 1   -- really too bold
 -- local bold_strength_factor = 1/2 -- bold enough
 local bold_strength_factor = 3/8 -- as crengine, lighter
+local cjk_bold_strength_factor = 2/8 -- a bit less bold for complex CJK glyphs
+                -- or try 1/8

 -- Add some properties to a face object as needed
 local _completeFaceProperties = function(face_obj)
     if not face_obj.embolden_half_strength then
+        local factor = bold_strength_factor
+        if face_obj.realname == "NotoSansCJKsc-Regular.otf" then
+            factor = cjk_bold_strength_factor
+        end
         -- Cache this value in case we use bold, to avoid recomputation
-        face_obj.embolden_half_strength = face_obj.ftface:getEmboldenHalfStrength(bold_strength_factor)
+        face_obj.embolden_half_strength = face_obj.ftface:getEmboldenHalfStrength(factor)
     end
 end
```

**Comment by weijiuqiao** [comment]
I understand your suggestion to tune down the bold strength so that strokes are less likely to be blurred together. I tried the values and made some screenshots. 
![xxx](https://user-images.githubusercontent.com/59040746/174426979-177b6ff5-5294-438c-9478-b018e3092b88.png)

Yes when it's less thick, it's cleaner, but it also just makes the bold not bold, invalidating the reason to use bold in the first place, I guess? 

Also, despite the difference being subtle, I'm sure you are well aware that a bold face is not just thickening the strokes of a regular face. In case of the CJK font, a proper bold face should give uncompromised thickness without blurring the details, and that's not easy for a synthesizer to do. So in the end the real reason for this fr is more of aesthetics than practicality :) (well, using a font subset is being practical)

**Comment by weijiuqiao** [comment]
I understand your suggestion to tune down the bold strength factor to make it "less bold" so that strokes are less likely to blur together. I tried the values 1/8 and 2/8. They definitely worked in this regard but also just aren't that bold anymore.

Despite the difference being subtle, I'm sure you are quite aware that a bold typeface is much more than thickening the strokes of its regular counterpart. I'm not saying that's only what the synthesizer does, but the result really does not match a hand-crafted bold face. Otherwise we would just synthesize all fonts instead of having a separate bold face. 

A proper CJK bold font should give uncompromised real bold thickness that balances well with other scripts while keeping the details clean, that's not easy for the synthesizer. So in the end, this fr is more of aesthetics than practicality :) (but of course using a font subset is still being practical).

**Comment by poire-z** [comment]
Yes, I'm aware (and the synthetiser may not do more than thickening the strokes :).
But a cjk bold font subset, made from the chars that are used today in CJK translations, will be a maintenance burden (with the need to regenerate them on each new translation).
Also, it's not only our hardcoded translatable strings that may be rendered in bold: filenames and book titles (and possibly other metadata) too (ie. in Detailed list filemanage view) - and there, you may get chars not in your subset synthesized near other real chars, and the difference inside a single string may look even uglier.

**Comment by weijiuqiao** [comment]
> But a cjk bold font subset, made from the chars that are used today in CJK translations, will be a maintenance burden (with the need to regenerate them on each new translation).

Might be one of the standardized procedures during building the releases.

> Also, it's not only our hardcoded translatable strings that may be rendered in bold: filenames and book titles (and possibly other metadata) too (ie. in Detailed list filemanage view) - and there, you may get chars not in your subset synthesized near other real chars, and the difference inside a single string may look even uglier.

Well, this I can't object. Then it has to be something like a choice to download the complete bold face, like the dictionaries, for the users in need, or let them set a custom font as the bold face (which doesn't sound like the best idea).

**Comment by poire-z** [comment]
> for the users in need

"in need" vs luxury :)
I'd say lower the embolden factor for a list of fonts that would need it, so users aren't "in need".
Then, may be add something to pick and associate a Bold font if one is found (a bit like #8722, but without any UI) - so aesthets can do the manual work of downloading it, put i on their device, and it would just be picked without any UI or interaction need.

If you want to go with a font downloading infrastructure, please think about allowing downloading Noto Sans versions for all the languages we don't ship one for (+ their bold or italic variants as an option) - so they benefit from the fallback trick of #8722.

**Comment by weijiuqiao** [comment]
>may be add something to pick and associate a Bold font if one is found (a bit like https://github.com/koreader/koreader/pull/8722, but without any UI) - so aesthets can do the manual work of downloading it, put i on their device, and it would just be picked without any UI or interaction need.

I'm most pro this solution, loosening the limit on additional fallback to include bold font as well. So the about text would also need to update a bit to something like "regular not italic nor serif will appear in this menu, and bold will be used if exists"

**Comment by poire-z** [comment]
I don't think any bold font should appear among that menu items (but you can mention a thing about bold in the help_text).
I think you would just need to have any found bold font having a (shipped or user added) existing regular font added in here:
https://github.com/koreader/koreader/blob/83453dfd8f718d022783ad3182d068e9a9f28ddd/frontend/ui/font.lua#L10-L22
so you end up having added:
`_bold_font_variant["NotoSansCJKsc-Regular.otf"] = "NotoSansCJKsc-Bold.otf" `

**Comment by weijiuqiao** [comment]
Or maybe a little bit UI in the list of font names, so a user could know if it succeeded? Something like
    with bold: Noto Sans XXX (Regular + Bold)
    without bold: Noto Sans XXX (Regular)
or
    with bold: Noto Sans XXX 
    without bold: Noto Sans XXX (Regular)
or
    with bold: Noto Sans XXX 
    without bold: Noto Sans XXX (auto Bold)
or
    with bold: Noto Sans XXX (with Bold)
    without bold: Noto Sans XXX
?

**Comment by poire-z** [comment]
Well, I guess he'll be able to tell if he felt the need to add a bold font because synthesized bold sucked :)
Our font list is CRE specific, I don't think it needs to add noise to indicate the variants (ital, bold, boldital) available for a font.
At most some logger.info() that it was found and added.
